sycl/source/detail/plugin.hpp

```diff
@@ -184,11 +184,71 @@ class plugin {
   void *getLibraryHandle() { return MLibraryHandle; }
   int unload() { return RT::unloadPlugin(MLibraryHandle); }
 
+  // return the index of PiPlatforms.
+  // If not found, add it and return its index.
+  int getPlatformId(RT::PiPlatform Platform) {
+    if (PiPlatforms) {
+      auto It = std::find(PiPlatforms->begin(), PiPlatforms->end(), Platform);
+      if (It != PiPlatforms->end())
+        return It - PiPlatforms->begin();
+
+      PiPlatforms->push_back(Platform);
+      LastDeviceIds.push_back(0);
+      return PiPlatforms->size() - 1;
+    }
+
+    return -1;
+  }
+  // Device ids are consecutive across platforms within a plugin.
+  // We need to return the same starting index for the given platform.
+  // So, instead of returing the last device id of the given platform,
+  // return the last device id of the predecessor platform.
+  int getStartingDeviceId(RT::PiPlatform Platform) {
+    if (!PiPlatforms)
+      PiPlatforms = new std::vector<RT::PiPlatform>;
+    int PlatformId = getPlatformId(Platform);
+    if (PlatformId == 0)
+      return 0;
+
+    return LastDeviceIds[PlatformId - 1];
+  }
+  // set the id of the last device for the given platform
+  void setLastDeviceId(RT::PiPlatform Platform, int Id) {
+    int PlatformId = getPlatformId(Platform);
+    LastDeviceIds[PlatformId] = Id;
+  }
+  // reset all last device ids to zeros
+  void resetLastDeviceIds() {
+    std::fill(LastDeviceIds.begin(), LastDeviceIds.end(), 0);
+  }
+
+  // Do not leave PiPlatform pointers so that they will not be
+  // released twice at the shutdown time.
+  void resetPiPlatforms() {
+    delete PiPlatforms;
+    PiPlatforms = nullptr;
+  }
```

Comment:
Could you please clarify why `PiPlatform` pointers can be released twice?
BTW. Can we use `std::unique_ptr` for `PiPlatforms`?

Reply:
plugins maintain its own cache of PiPlatforms and they are destroyed at the shutdown time.
If we leave the vector of PiPlatforms in the plugin, the plugin destructor will call destructor of PiPlatforms and they will cause multiple destruction.

Reply:
I'm not sure I understood.  If `PiPlatforms` was plain `std::vector` how it would be destroyed multiple times?

Reply:
std::vector<EltType> destructor calls destructors of each element.
Since plugin contains std::vector<PiPlatform>, it will call destructors of each PiPlatform.

Reply:
It seems there are still races between: 
`  int DeviceNum = Plugin.getStartingDeviceId(Platform);`
`  Plugin.setLastDeviceId(Platform, DeviceNum);`
`    Plugins[i].resetLastDeviceIds();`
`        Plugins[i].getPlatformId(PiPlatform);`